climada/trajectories/impact_calc_strat.py

```diff
@@ -0,0 +1,160 @@
+        ).impact()
+        return imp_E0H0, imp_E1H0, imp_E0H1, imp_E1H1
+
+    def _apply_risk_transfer(
+        self,
+        impacts: tuple[Impact, Impact, Impact, Impact],
+        risk_transf_attach: float | None,
+        risk_transf_cover: float | None,
+        calc_residual: bool,
+    ):
+        """Apply risk transfer to the calculated impacts."""
+        if risk_transf_attach is not None and risk_transf_cover is not None:
+            for imp in impacts:
+                imp.imp_mat = self.calculate_residual_or_risk_transfer_impact_matrix(
+                    imp.imp_mat, risk_transf_attach, risk_transf_cover, calc_residual
+                )
+
+    def calculate_residual_or_risk_transfer_impact_matrix(
+        self, imp_mat, risk_transf_attach, risk_transf_cover, calc_residual
+    ):
+        """
+        Calculate either the residual or the risk transfer impact matrix.
+
+        The impact matrix is adjusted based on the total impact for each event.
+        When calculating the residual impact, the result is the total impact minus
+        the risk layer. The risk layer is defined as the minimum of the cover and
+        the maximum of the difference between the total impact and the attachment.
+        If `calc_residual` is False, the function returns the risk layer matrix
+        instead of the residual.
+
+        Parameters
+        ----------
+        imp_mat : scipy.sparse.csr_matrix
+            The original impact matrix to be scaled.
+        attachment : float, optional
+            The attachment point for the risk layer.
+        cover : float, optional
+            The maximum coverage for the risk layer.
+        calc_residual : bool, default=True
+            Determines if the function calculates the residual (if True) or the
+            risk layer (if False).
+
+        Returns
+        -------
+        scipy.sparse.csr_matrix
+            The adjusted impact matrix, either residual or risk transfer.
+
+        """
+
+        if risk_transf_attach and risk_transf_cover:
+            imp_mat = copy.deepcopy(imp_mat)
+            # Calculate the total impact per event
+            total_at_event = imp_mat.sum(axis=1).A1
+            # Risk layer at event
+            transfer_at_event = np.minimum(
+                np.maximum(total_at_event - risk_transf_attach, 0), risk_transf_cover
+            )
+            residual_at_event = np.maximum(total_at_event - transfer_at_event, 0)
+
+            # Calculate either the residual or transfer impact matrix
+            # Choose the denominator to rescale the impact values
+            if calc_residual:
+                numerator = residual_at_event
+            else:
+                numerator = transfer_at_event
+
+            rescale_impact_values = np.divide(
+                numerator,
+                total_at_event,
+                out=np.zeros_like(numerator, dtype=float),
+                where=total_at_event != 0,
+            )
```

Comment:
Does this mean that the insurance cover is shared among all exposure points and equally distributed? If yes, I think this needs to be made very clear in the docstring. Otherwise I am confused about what this relative residuals / transfers mean.